src/js/all/stories/typescript/ExtendTheme.tsx

```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+import { css } from 'styled-components';
+import { storiesOf } from '@storybook/react';
+import { ThemeType, BoxTypes, Grommet, Box, Anchor, Button } from 'grommet';
+
+// Custom theme to verify that various `extend` types work correctly
+const custom: ThemeType = {
+  box: {
+    extend: css`
+      cursor: ${(props: BoxTypes) => (props.onClick ? 'cursor' : 'inherit')};
+    `,
+  },
+  anchor: {
+    extend: props => css`
+      color: ${props.href ? 'green' : 'red'};
+    `,
+  },
+  button: {
+    extend: props => {
+      let extraStyles = '';
+      if (props.primary) {
+        extraStyles = `
+          text-transform: uppercase;
+          color: white;
+        `;
+      }
+      return `
+        color: green;
+        font-size: 12px;
+        font-weight: bold;
+
+        ${extraStyles}
+      `;
+    },
+  },
+  icon: {
+    // Components without typed theme props should pass type-checking
+    extend: props => css`
+      color: ${props.untypedProp === 'some value' ? 'red' : 'green'};
+    `,
+  },
+};
+
+const ExtendTheme: React.FC = () => {
+  return (
+    <Grommet theme={custom}>
+      <Box pad="small" gap="medium" width="medium">
+        <Anchor href="#">Anchor (href)</Anchor>
+        <Anchor>Anchor (no href)</Anchor>
+
+        <Button label="Standard Button" onClick={() => {}} />
+        <Button label="Primary Button" primary onClick={() => {}} />
```

Comment:
This was to test that we can define extra styles and that it passes type checking. I wanted to have a few working examples rather than just have a theme definition alone, so figured Buttons were a good test of that.